### DISCUSSION TITLE
USBSID-Pico, a new hardware SID player and a question :)

### DISCUSSION THREAD

**Discussion by LouDnl** [libsidplayfp/libsidplayfp#171] [Show and tell]
Hi all,

I am the creator of [USBSID-Pico](https://github.com/LouDnl/USBSID-Pico) and have started development on a driver for and an implementation in libsidplayfp and sidplayfp. When finished I plan to create pull-requests for both repo's.
The board has 2 sockets and supports real SID chips and hardware clones. When using hardware clones you can use up to 4x SID.

_Now for my question:_
At the moment I am having trouble with finding a correct solution to play multisid tunes. I have been roaming around the codebase to figure out how to implement working support for this, but at the moment I haven't gotten further then enabling 4 sids in the builder source and restricting the emu source to one actual enabled USB object. But whatever I try, the driver never receives the correct sid chip id.
Any tips would be very much appreciated and thanks in advance!

For example when using Vice, the hardware SID driver receives an extra argument that contains the SID chip number where the write needs to go to. This easilly translates to `n * 0x20` for the correct `uint8_t` address on the USBSID-Pico.
And with SidBerry the hardware SID driver receives a `uint16_t` address that is converted by the driver to the correct `uint8_t` address.
Even the `Hardsid/SidBlaster USB` emulation driver I made for use with `jsidplay2` receives a device id that I can translate to the correct address.

Want to try it out now? Forks with (working) crude first versions can be found here:
https://github.com/LouDnl/libsidplayfp-usbsid (be sure to add `--with-usbsid ` as argument to `./configure`
https://github.com/LouDnl/sidplayfp-usbsid
https://github.com/LouDnl/USBSID-Pico-driver

![image](https://github.com/user-attachments/assets/95205e4b-1766-49ea-85fa-2ae1c71e7873)

**Comment by mywave82** [comment]
If I understand your problem correctly, you need to differentiate which player SID chip should end up at which physical SID chip on your device.

Doing a super quick peek at your code, you could add a parameter to the constructor of libsidplayfp::USBSID object that makes each instance aware of its "position".

So change:
`std::unique_ptr<libsidplayfp::USBSID> sid(new libsidplayfp::USBSID(this, m_isthreaded));
`
into
`std::unique_ptr<libsidplayfp::USBSID> sid(new libsidplayfp::USBSID(this, m_isthreaded, count));
`
and store that count (which actually is the index) into your object and do your magic.

**Reply by LouDnl** [reply]
Thanks for taking time to have a look at this, will let you know if this works!

**Reply by LouDnl** [reply]
Okay, I tried this and of course added some debug logging, up to the device reset this seems to do something, but that reset is triggered from within the constructor.
The writes called cannot differentiate what sidno its apparently, all writes seem to head to the same object.
```bash
[USBSIDBuilder USBSID] m_initialised:1 m_count:4 m_status:1 m_isthreaded:0
[create]A sids:3 count:- m_status:1 m_count:4 m_initialised:1
[create]B sids:3 count:4 m_status:1 m_count:4 m_initialised:1
[USBSID] Driver init start
[USBSID USBSID 0]A sidno:0 sid:0 num:4 m_sidFree:0000 m_instance:0 m_status:0
[USBSID]A num:4 i:0 m_sidFree[i]:0 sid:0
[USBSID]B sid:0 num:0 m_sidFree:1000 m_instance:0 m_status:0
[USBSID] Setup start
[USBSID] Alloc out_buffer complete
[USBSID] Alloc transfer_out complete
[USBSID] libusb_fill_bulk_transfer transfer_out complete
[USBSID] Alloc in_buffer complete
[USBSID] Alloc transfer_in complete
[USBSID] libusb_fill_bulk_transfer transfer_in complete
[USBSID]C m_handle:0 sid:0 num:0 m_sidFree:1000 m_instance:1 m_status:0
[reset] sidno:0
[USBSID] Pause
[USBSID] Reset
[create]C -1841282056 sids:3 count:0 m_status:1 m_count:4 m_initialised:1
[USBSID] Driver init start
[USBSID USBSID 1]A sidno:1 sid:0 num:0 m_sidFree:1000 m_instance:1 m_status:0
[USBSID]A num:0 i:0 m_sidFree[i]:1 sid:0
[USBSID]B num:0 i:0 m_sidFree[i]:1 sid:0
[USBSID]A num:0 i:1 m_sidFree[i]:0 sid:0
[USBSID]B sid:1 num:1 m_sidFree:1100 m_instance:1 m_status:0
[USBSID]C m_handle:1 sid:1 num:1 m_sidFree:1100 m_instance:1 m_status:0
[reset] sidno:1
[USBSID] Pause
[USBSID] Reset
[create]C -1841282056 sids:3 count:1 m_status:1 m_count:4 m_initialised:1
[USBSID] Driver init start
[USBSID USBSID 2]A sidno:2 sid:1 num:1 m_sidFree:1100 m_instance:1 m_status:0
[USBSID]A num:1 i:0 m_sidFree[i]:1 sid:1
[USBSID]B num:1 i:0 m_sidFree[i]:1 sid:1
[USBSID]A num:1 i:1 m_sidFree[i]:1 sid:1
[USBSID]B num:1 i:1 m_sidFree[i]:1 sid:1
[USBSID]A num:1 i:2 m_sidFree[i]:0 sid:1
[USBSID]B sid:2 num:2 m_sidFree:1110 m_instance:1 m_status:0
[USBSID]C m_handle:2 sid:2 num:2 m_sidFree:1110 m_instance:1 m_status:0
[reset] sidno:2
[USBSID] Pause
[USBSID] Reset
[create]C -1841282056 sids:3 count:2 m_status:1 m_count:4 m_initialised:1
[create]D sids:3 count:3 m_status:1 m_count:4 m_initialised:1
[USBSID] Set clockspeed to: [i]985248 ~ [r]985248
[sampling] m_handle:0 m_handle:0 sid:2 num:2 m_instance:1 m_status:1 m_sidFree:1110
[USBSID] Set clockspeed to: [i]985248 ~ [r]985248
[sampling] m_handle:0 m_handle:0 sid:2 num:2 m_instance:1 m_status:1 m_sidFree:1110
[reset] sidno:0
[USBSID] Pause
[USBSID] Reset
[reset] sidno:0
[USBSID] Pause
[USBSID] Reset
```

**Comment by LouDnl** [comment]
From what I can see inside the libsidplayfp code is that the address gets changed from an unsigned 16bit integer to an unsigned 8 bit integer here: https://github.com/libsidplayfp/libsidplayfp/blob/51158a807d7a8f748a7aa62a8519e9878e1a5cd5/src/c64/c64sid.h#L60-L64

This shouldn't be an issue if the correct SID object received the correct write, but for some reason I am missing why this isn't happening. There are actually 3 objects being created if my logging is correct that is.

**Comment by LouDnl** [comment]
Okay, so I've fixed it and multi SID play is now working.
What I didn't know is that this part actually locks the first object into place and stops the other objects from being reached, no clue why though. I used the hardsid and exsid builder code as example and edited it for usbsid. When I saw that the resid and residfp did not use that piece of code it brought me to getting it to work.
https://github.com/LouDnl/libsidplayfp-usbsid/blob/919ee5a3a811d4d6373d6d85803b24a4cd17e714/src/builders/usbsid-builder/usbsid-emu.cpp#L228-L240

**Comment by LouDnl** [comment]
Closed with https://github.com/libsidplayfp/libsidplayfp/pull/216
